src/wamr/env.cpp

```diff
@@ -0,0 +1,97 @@
+#include <faabric/util/logging.h>
+#include <storage/FileDescriptor.h>
+#include <wamr/WAMRWasmModule.h>
+#include <wamr/native.h>
+
+#include <stdexcept>
+
+// WAMR includes
+#include <wasm_export.h>
+#include <wasmtime_ssp.h>
+
+namespace wasm {
+uint32_t wasi_args_get(wasm_exec_env_t exec_env,
+                       uint32_t* argvOffsetsWasm,
+                       char* argvBuffWasm)
+{
+    SPDLOG_DEBUG("S - args_get");
+
+    WAMRWasmModule* module = getExecutingWAMRModule();
+    module->writeArgvToWamrMemory(argvOffsetsWasm, argvBuffWasm);
+
+    return __WASI_ESUCCESS;
+}
+
+uint32_t wasi_args_sizes_get(wasm_exec_env_t exec_env,
+                             uint32_t* argcWasm,
+                             uint32_t* argvBuffSizeWasm)
+{
+    SPDLOG_DEBUG("S - args_sizes_get");
+
+    WAMRWasmModule* module = getExecutingWAMRModule();
+
+    module->validateNativeAddress(argcWasm, sizeof(uint32_t));
+    module->validateNativeAddress(argvBuffSizeWasm, sizeof(uint32_t));
+
+    *argcWasm = module->getArgc();
+    *argvBuffSizeWasm = module->getArgvBufferSize();
+
+    return __WASI_ESUCCESS;
+}
+
+uint32_t wasi_environ_get(wasm_exec_env_t exec_env,
+                          uint32_t* envOffsetsWasm,
+                          char* envBuffWasm)
+{
+    SPDLOG_DEBUG("S - environ_get");
+
+    WAMRWasmModule* module = getExecutingWAMRModule();
+    module->writeWasmEnvToWamrMemory(envOffsetsWasm, envBuffWasm);
+
+    return __WASI_ESUCCESS;
+}
+
+uint32_t wasi_environ_sizes_get(wasm_exec_env_t exec_env,
+                                int32_t* envCountWasm,
+                                int32_t* envBufferSizeWasm)
+{
+    SPDLOG_DEBUG("S - environ_sizes_get");
+
+    WAMRWasmModule* module = getExecutingWAMRModule();
+    WasmEnvironment& wasmEnv = module->getWasmEnvironment();
+
+    module->validateNativeAddress(envCountWasm, sizeof(uint32_t));
+    module->validateNativeAddress(envBufferSizeWasm, sizeof(uint32_t));
+
+    *envCountWasm = wasmEnv.getEnvCount();
+    *envBufferSizeWasm = wasmEnv.getEnvBufferSize();
+
+    return __WASI_ESUCCESS;
+}
+
+void wasi_proc_exit(wasm_exec_env_t exec_env, int32_t retCode)
+{
+    SPDLOG_DEBUG("S - proc_exit {}", retCode);
+
+    // 08/09/2021 - WAMR doesn't support throwing and catching exceptions.
+    // Alternatively, set the error message.
+    WAMRWasmModule* module = getExecutingWAMRModule();
+    char buf[32];
```

Comment:
It's not clear where this `32` comes from, is it `sizeof(int)`? If so, can use that instead.